File: apps/agent/src/lib/critique.ts
import { ObjectId, type Collection } from "mongodb";
import { getDb } from "./mongo.js";

export type FindingSeverity = "high" | "medium" | "low";

export interface CritiqueFinding {
  severity: FindingSeverity;
  claim: string;          // quoted from the draft
  issue: string;          // what's wrong
  evidence: "supported" | "unsupported" | "contradicted" | "missing";
  citation?: string;      // chunk id or title that informs the verdict
  suggestion?: string;    // concrete revision
}

export interface CritiqueRecord {
  _id?: ObjectId;
  actionId: string;            // string form of ActionRecord._id
  runId?: string;
  query?: string;
  verdict: "approve" | "revise" | "reject";
  summary: string;             // one-sentence overall read
  findings: CritiqueFinding[];
  voice: { score: number; notes: string };  // 0–10 + remarks
  model?: string;
  createdAt: Date;
}

export async function critiquesCol(): Promise<Collection<CritiqueRecord>> {
  const db = await getDb();
  return db.collection<CritiqueRecord>("critiques");
}

export async function saveCritique(rec: Omit<CritiqueRecord, "_id" | "createdAt">): Promise<string> {
  const col = await critiquesCol();
  const doc: CritiqueRecord = { ...rec, createdAt: new Date() };
  const ins = await col.insertOne(doc);
  return ins.insertedId.toString();
}

export async function getCritique(id: string): Promise<CritiqueRecord | null> {
  if (!ObjectId.isValid(id)) return null;
  const col = await critiquesCol();
  return col.findOne({ _id: new ObjectId(id) });
}

export async function getCritiqueByAction(actionId: string): Promise<CritiqueRecord | null> {
  const col = await critiquesCol();
  return col.findOne({ actionId }, { sort: { createdAt: -1 } });
}

export function publicCritique(c: CritiqueRecord): Record<string, unknown> {
  return {
    id: c._id?.toString() ?? "",
    actionId: c.actionId,
    runId: c.runId ?? null,
    query: c.query ?? null,
    verdict: c.verdict,
    summary: c.summary,
    findings: c.findings,
    voice: c.voice,
    model: c.model ?? null,
    createdAt: c.createdAt instanceof Date ? c.createdAt.toISOString() : c.createdAt,
  };
}

File: apps/agent-py/app/lib/critique.py
"""Critique store — port of apps/agent/src/lib/critique.ts."""
from __future__ import annotations

from datetime import datetime, timezone

from bson import ObjectId

from app.db.mongo import collection


def critiques_col():
    return collection("critiques")


async def save_critique(rec: dict) -> str:
    doc = {**rec, "createdAt": datetime.now(timezone.utc)}
    ins = await critiques_col().insert_one(doc)
    return str(ins.inserted_id)


async def get_critique(cid: str) -> dict | None:
    if not ObjectId.is_valid(cid):
        return None
    return await critiques_col().find_one({"_id": ObjectId(cid)})


async def get_critique_by_action(action_id: str) -> dict | None:
    return await critiques_col().find_one({"actionId": action_id}, sort=[("createdAt", -1)])


def _iso(v):
    return v.isoformat() if isinstance(v, datetime) else v


def public_critique(c: dict) -> dict:
    return {
        "id": str(c.get("_id", "")), "actionId": c["actionId"], "runId": c.get("runId"),
        "query": c.get("query"), "verdict": c["verdict"], "summary": c["summary"],
        "findings": c["findings"], "voice": c["voice"], "model": c.get("model"),
        "createdAt": _iso(c.get("createdAt")),
    }
